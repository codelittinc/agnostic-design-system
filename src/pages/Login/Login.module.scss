@import '~@codelitt/ay-css-library/src/components/brand/colors';

.login-container {
  width: 22rem;
  margin: 0 auto;
}

.logo-container {
  margin: 0;
  margin-bottom: 5rem;
  padding-top: 2rem;
  text-align: center;
}

.login-form {
  .login-input {
    margin-bottom: 1rem;
    width: 100%;
  }

  .sign-me-in-button {
    margin-top: 1rem;
    width: 100%;
  }
}