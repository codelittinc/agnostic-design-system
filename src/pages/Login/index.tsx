import React, { useState } from 'react';
import { RouteComponentProps } from 'react-router';

import styles from './Login.module.scss';
import Button from '../../components/Button';
import Col from '../../components/Grid/Col';
import Container from '../../components/Grid/Container';
import Input from '../../components/Input';
import Row from '../../components/Grid/Row';
import Text from '../../components/Text';

import LogoImg from '../../assets/images/logo-90x48.png';
import { authContext } from './contexts/AuthContext';

type RouteProps = RouteComponentProps;

const Login: React.FC<RouteProps> = (props: RouteProps) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [errors, setErrors] = useState({ email: '', password: '' });
  const auth = React.useContext(authContext);

  const validateLoginForm = (): boolean => {
    const user = 'user@ay.com';
    const pass = '1234';

    if (user !== email) {
      setErrors({ ...errors, email: 'Email address is incorrect.' });
      return false;
    }

    if (pass !== password) {
      setErrors({ ...errors, password: 'Password is incorrect.' });
      return false;
    }

    auth.setAuth({ id: 1, email });
    return true;
  };

  const isEmail = (emailAddress: string) => {
    const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(String(emailAddress).toLowerCase());
  };

  const isValidEmail = () => isEmail(email);

  const validateEmail = (value: string) => {
    setEmail(value);

    if (!isEmail(value)) {
      setErrors({ ...errors, email: 'Please enter a valid email address.' });
    } else {
      setErrors({ ...errors, email: '' });
    }
  };

  const isSubmitDisabled = () => {
    return !email || !isValidEmail() || !password;
  };

  const login = () => {
    const path = props.location.state.from.pathname;
    if (validateLoginForm()) {
      props.history.push(path);
    }
  };

  return (
    <div className={styles['login-background']}>
      <Container>
        <Row wrapperClassName={styles['logo-container']}>
          <Col>
            <img className={styles.logo} src={LogoImg} alt="Avison Young" />
          </Col>
        </Row>

        <Row wrapperClassName={styles['login-form']}>
          <Col>
            <Container>
              <Row wrapperClassName={styles.header}>
                <Col>
                  <Text
                    wrapperClassName={styles['header-text']}
                    type="heading"
                    headingSize={5}
                    value="Nice to see you"
                  />
                  <Text
                    wrapperClassName={styles['header-text']}
                    value="Please, enter your credentials"
                  />
                </Col>
              </Row>

              <Row>
                <Col>
                  <form>
                    <Input
                      type="text"
                      placeholder="Email"
                      status={errors.email ? 'error' : undefined}
                      statusText={errors.email}
                      value={email}
                      onChange={validateEmail}
                      wrapperClassName={styles['login-input']}
                    />

                    <Input
                      type="password"
                      placeholder="Password"
                      status={errors.password ? 'error' : undefined}
                      statusText={errors.password}
                      value={password}
                      onChange={setPassword}
                      wrapperClassName={styles['login-input']}
                    />

                    <Button
                      wrapperClassName={styles['sign-me-in-button']}
                      iconPosition="right"
                      isDisabled={isSubmitDisabled()}
                      label="Sign me in"
                      onClick={login}
                      size="small"
                      type="main"
                    />
                  </form>
                </Col>
              </Row>
            </Container>
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default Login;
