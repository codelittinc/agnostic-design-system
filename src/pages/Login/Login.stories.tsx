import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import { createMemoryHistory, createLocation } from 'history';
import Login from '.';

type UserAuth = {
  id: number;
  email: string;
};

const stories = storiesOf('Pages|Login', module);
const history = createMemoryHistory();
const path = `/route/:id`;

const match = {
  isExact: false,
  path,
  url: path.replace(':id', '1'),
  params: { id: '1' },
};

const location = createLocation(match.url);
location.state = { from: { pathname: '/' } };

const callback = (user: UserAuth) => {
  alert(`Logged in as ${user.email}`);
};

stories.add(
  'Login Page',
  () => (
    <div>
      <Login
        routeProps={{
          history: { ...history },
          location: { ...location },
          match: { ...match },
        }}
        callback={callback}
      />
    </div>
  ),
  {
    order: 1,
    info: `
    The \`Login\` component is designed to work with react router routes.  It validates the user and executes a callback function with a parameter of the following type:

    type UserAuth = {
      id: number;
      email: string;
    };

    The \`Login\` component use the \`history\`, \`location\` and \`match\` props from React Router.  These props are used to handle the browse history,
    and it uses the location to redirect the user to a specific page after the user is authenticated.

    Once the user is validated, it executes the callback function provided by the caller. 

    Usage instructions:
    
        import { Login as LoginComponent, UserAuth } from '@codelitt/ay-design-library';

        const onValidate = (user: UserAuth) => {
          // handle the user after it's validated.  In this case, it's being set in a context
          auth.setAuth({ id: user.id, email: user.email });
        };
    
        <LoginComponent
        routeProps={{
          history: { ...history },
          location: { ...location },
          match: { ...match },
        }}
        callback={onValidate}
      />
        `,
  },
);
