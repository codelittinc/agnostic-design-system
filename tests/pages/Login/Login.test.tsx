import React from 'react';
import { render } from '@testing-library/react';
import { createMemoryHistory, createLocation } from 'history';
import Login from '../../../src/pages/Login';

describe('Login', () => {
  const getComponent = (callback: () => void) => {
    const history = createMemoryHistory();
    const path = `/route/:id`;

    const match = {
      isExact: false,
      path,
      url: path.replace(':id', '1'),
      params: { id: '1' },
    };

    const location = createLocation(match.url);

    return render(
      <Login
        routeProps={{
          history: { ...history },
          location: { ...location },
          match: { ...match },
        }}
        callback={callback}
      />,
    );
  };
  it('renders without crashing', () => {
    const callback = () => {};
    const login = getComponent(callback);
    expect(login).toMatchSnapshot();
  });
});
