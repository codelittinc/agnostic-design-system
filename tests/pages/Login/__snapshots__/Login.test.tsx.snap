// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-background"
      >
        <div
          class="ay-container"
        >
          <div
            class="ay-row logo-container"
          >
            <div
              class="ay-col-xs-4"
            >
              <img
                alt="Avison Young"
                class="logo"
                src=""
              />
            </div>
          </div>
          <div
            class="ay-row login-form"
          >
            <div
              class="ay-col-xs-4"
            >
              <div
                class="ay-container"
              >
                <div
                  class="ay-row header"
                >
                  <div
                    class="ay-col-xs-4"
                  >
                    <h5
                      class="heading header-text"
                    >
                      Nice to see you
                    </h5>
                    <span
                      class="body-content header-text"
                    >
                      Please, enter your credentials
                    </span>
                  </div>
                </div>
                <div
                  class="ay-row"
                >
                  <div
                    class="ay-col-xs-4"
                  >
                    <form>
                      <div
                        class="wrapper login-input"
                      >
                        <div
                          class="input-container"
                        >
                          <input
                            class="input undefined-status"
                            placeholder="Email"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="wrapper login-input"
                      >
                        <div
                          class="input-container"
                        >
                          <div
                            class="container"
                          >
                            <input
                              class="input undefined-status"
                              placeholder="Password"
                              type="password"
                              value=""
                            />
                            <button
                              aria-label="Reveal Password"
                              class="button password"
                              id="btn-undefined"
                              type="button"
                            />
                          </div>
                        </div>
                      </div>
                      <button
                        class="button main right-icon small-size sign-me-in-button"
                        disabled=""
                        type="button"
                      >
                        Sign me in
                      </button>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-background"
    >
      <div
        class="ay-container"
      >
        <div
          class="ay-row logo-container"
        >
          <div
            class="ay-col-xs-4"
          >
            <img
              alt="Avison Young"
              class="logo"
              src=""
            />
          </div>
        </div>
        <div
          class="ay-row login-form"
        >
          <div
            class="ay-col-xs-4"
          >
            <div
              class="ay-container"
            >
              <div
                class="ay-row header"
              >
                <div
                  class="ay-col-xs-4"
                >
                  <h5
                    class="heading header-text"
                  >
                    Nice to see you
                  </h5>
                  <span
                    class="body-content header-text"
                  >
                    Please, enter your credentials
                  </span>
                </div>
              </div>
              <div
                class="ay-row"
              >
                <div
                  class="ay-col-xs-4"
                >
                  <form>
                    <div
                      class="wrapper login-input"
                    >
                      <div
                        class="input-container"
                      >
                        <input
                          class="input undefined-status"
                          placeholder="Email"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class="wrapper login-input"
                    >
                      <div
                        class="input-container"
                      >
                        <div
                          class="container"
                        >
                          <input
                            class="input undefined-status"
                            placeholder="Password"
                            type="password"
                            value=""
                          />
                          <button
                            aria-label="Reveal Password"
                            class="button password"
                            id="btn-undefined"
                            type="button"
                          />
                        </div>
                      </div>
                    </div>
                    <button
                      class="button main right-icon small-size sign-me-in-button"
                      disabled=""
                      type="button"
                    >
                      Sign me in
                    </button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
